### PR TITLE
Add shutdown room API

### PR DESCRIPTION
**Pull request by erikjohnston**
Tested by: https://github.com/matrix-org/sytest/pull/360

### CODE REVIEW

File: synapse/handlers/room.py [hunk 1]
```diff
@@ -316,6 +327,8 @@ def send(etype, content, **kwargs):
                 for invitee in invite_list:
                     power_level_content["users"][invitee] = 100
 
+            power_level_content.update(power_level_content_override)
```
Comment (NegativeMjark) [inline synapse/handlers/room.py]: Did you mean for the power_level_content to end up in this PR?

Comment (erikjohnston) [inline synapse/handlers/room.py]: I guess it would have been nice to split the PR in two.

**Comment by NegativeMjark** [comment]
Do we also need to worry about the calls to send_membership_event?

https://github.com/matrix-org/synapse/blob/36e51aad3c5f2477bdfe0ca77e09e40d7c9b6291/synapse/handlers/federation.py#L2065

https://github.com/matrix-org/synapse/blob/36e51aad3c5f2477bdfe0ca77e09e40d7c9b6291/synapse/handlers/federation.py#L2099

File: synapse/handlers/room.py (hunk 1 above)
Comment (NegativeMjark) [inline synapse/handlers/room.py]: Ah, I see that you are using the overrides as part of the shutdown API.
The override stuff is trivial enough for me to to review as part of this PR.
It just looked orthogonal to me at first, hence me asking whether you meant to push it as part of the PR :)

**Comment by erikjohnston** [comment]
> Do we also need to worry about the calls to send_membership_event?

cat_fuck

File: synapse/storage/room.py [hunk 2]
```diff
@@ -507,3 +507,27 @@ def get_ratelimit_for_user(self, user_id):
...
             defer.returnValue(None)
+
+    @cached(max_entries=10000)
+    def is_room_blocked(self, room_id):
+        return self._simple_select_one_onecol(
+            table="blocked_rooms",
+            keyvalues={
+                "room_id": room_id,
+            },
+            retcol="1",
```
Comment (NegativeMjark) [inline synapse/storage/room.py]: This is slightly evil, I suspect that people looking at the _simple_select_one_onecol wouldn't expect someone to pass "1" as a column ID. But it's probably fine for now. :(

File: synapse/rest/client/v1/admin.py [hunk 3]
```diff
@@ -157,6 +158,118 @@ def on_POST(self, request, target_user_id):
...
+    """Shuts down a room by removing all local users from the room and blocking
+    all future invites and joins to the room. Any local aliases will be repointed
+    to a new room created by `new_room_user_id` and kicked users will be auto
+    joined to the new room.
+    """
+    PATTERNS = client_path_patterns("/admin/shutdown_room/(?P<room_id>[^/]+)")
+
+    DEFAULT_MESSAGE = (
+        "Sharing illegal content on this server is not permitted and rooms in"
+        " violatation will be blocked."
```
Comment (ara4n) [inline synapse/rest/client/v1/admin.py]: "violation"

Comment (turt2live) [inline synapse/rest/client/v1/admin.py]: fwiw, later on in this file is `This need a user have a administrator access in Synapse`.

Comment (ara4n) [inline synapse/rest/client/v1/admin.py]: fixed now.

**Pull request merged**